File: scripts/remove-tags.py
#!/usr/bin/env python3
import os
import re
from collections import defaultdict

content_dir = '/Users/masashimatsubuchi/my-first-project/content'

def find_md_files(directory):
    files = []
    for root, dirs, filenames in os.walk(directory):
        for filename in filenames:
            if filename.endswith('.md'):
                files.append(os.path.join(root, filename))
    return sorted(files)

def parse_tags_from_content(content):
    # Block format: tags:\n  - tag1\n  - tag2
    block_match = re.search(r'^tags:\s*\n((?:  - .+\n?)+)', content, re.MULTILINE)
    if block_match:
        lines = re.findall(r'  - (.+)', block_match.group(1))
        return 'block', [t.strip() for t in lines]
    # Inline format: tags: ["a", "b"] or tags: ['a', 'b']
    inline_match = re.search(r'^tags:\s*\[(.+)\]', content, re.MULTILINE)
    if inline_match:
        raw = inline_match.group(1)
        items = [s.strip().strip('"\'') for s in raw.split(',')]
        return 'inline', [t for t in items if t]
    return 'none', []

md_files = find_md_files(content_dir)
print(f"Found {len(md_files)} markdown files")

# Count tags
tag_count = defaultdict(int)
for filepath in md_files:
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    _, tags = parse_tags_from_content(content)
    for tag in tags:
        if tag:
            tag_count[tag] += 1

# Build remove set
always_remove = {'観葉植物', '図鑑'}
single_use = {tag for tag, count in tag_count.items() if count == 1}
remove_set = always_remove | single_use

print(f"Always remove: {always_remove}")
print(f"Single-use tags: {len(single_use)}")
print(f"Total tags to remove: {len(remove_set)}")

# Process files
modified = 0
for filepath in md_files:
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    fmt, tags = parse_tags_from_content(content)
    if fmt == 'none' or not tags:
        continue

    filtered = [t for t in tags if t not in remove_set]
    if len(filtered) == len(tags):
        continue

    removed = [t for t in tags if t in remove_set]
    rel = filepath.replace(content_dir + '/', '')

    if fmt == 'block':
        if filtered:
            new_block = 'tags:\n' + '\n'.join(f'  - {t}' for t in filtered) + '\n'
        else:
            new_block = 'tags: []\n'
        new_content = re.sub(r'^tags:\s*\n(?:  - .+\n?)+', new_block, content, flags=re.MULTILINE)
    elif fmt == 'inline':
        if filtered:
            new_inline = 'tags: [' + ', '.join(f'"{t}"' for t in filtered) + ']'
        else:
            new_inline = 'tags: []'
        new_content = re.sub(r'^tags:\s*\[.+\]', new_inline, content, flags=re.MULTILINE)

    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(new_content)

    modified += 1
    print(f"  {rel}: removed [{', '.join(removed)}]")

print(f"\nDone. Modified {modified} files.")

File: scripts/remove-tags.mjs
import { readFileSync, writeFileSync, readdirSync, statSync } from 'fs';
import { join } from 'path';

const contentDir = '/Users/masashimatsubuchi/my-first-project/content';

// Find all .md files
function findMdFiles(dir) {
  const files = [];
  for (const name of readdirSync(dir)) {
    const full = join(dir, name);
    if (statSync(full).isDirectory()) {
      files.push(...findMdFiles(full));
    } else if (name.endsWith('.md')) {
      files.push(full);
    }
  }
  return files;
}

const mdFiles = findMdFiles(contentDir);

// Parse tags from frontmatter
function parseTags(content) {
  const tags = [];
  // Block format: lines like "  - タグ" after "tags:"
  const blockMatch = content.match(/^tags:\s*\n((?:  - .+\n?)+)/m);
  if (blockMatch) {
    const lines = blockMatch[1].match(/  - (.+)/g) || [];
    for (const line of lines) {
      tags.push(line.replace(/^  - /, '').trim());
    }
    return { format: 'block', tags };
  }
  // Inline format: tags: ["a", "b"]
  const inlineMatch = content.match(/^tags:\s*\[(.+)\]/m);
  if (inlineMatch) {
    const raw = inlineMatch[1];
    const items = raw.split(',').map(s => s.trim().replace(/^["']|["']$/g, '').trim());
    return { format: 'inline', tags: items };
  }
  return { format: 'none', tags: [] };
}

// Count tag occurrences across all files
const tagCount = {};
for (const file of mdFiles) {
  const content = readFileSync(file, 'utf-8');
  const { tags } = parseTags(content);
  for (const tag of tags) {
    if (!tag) continue;
    tagCount[tag] = (tagCount[tag] || 0) + 1;
  }
}

// Build set of tags to remove
const alwaysRemove = new Set(['観葉植物', '図鑑']);
const singleUseTags = new Set(Object.entries(tagCount).filter(([, c]) => c === 1).map(([t]) => t));
const removeSet = new Set([...alwaysRemove, ...singleUseTags]);

console.log(`Total unique tags: ${Object.keys(tagCount).length}`);
console.log(`Always remove: ${[...alwaysRemove].join(', ')}`);
console.log(`Single-use tags to remove: ${singleUseTags.size}`);
console.log(`Total tags to remove: ${removeSet.size}`);

// Process files
let modifiedCount = 0;
for (const file of mdFiles) {
  let content = readFileSync(file, 'utf-8');
  const { format, tags } = parseTags(content);
  if (format === 'none' || tags.length === 0) continue;

  const filteredTags = tags.filter(t => !removeSet.has(t));
  if (filteredTags.length === tags.length) continue; // nothing removed

  // Reconstruct
  if (format === 'block') {
    // Find the tags block and replace
    const newBlock = filteredTags.length > 0
      ? `tags:\n${filteredTags.map(t => `  - ${t}`).join('\n')}\n`
      : `tags: []\n`;
    content = content.replace(/^tags:\s*\n(?:  - .+\n?)+/m, newBlock);
  } else if (format === 'inline') {
    const newInline = filteredTags.length > 0
      ? `tags: [${filteredTags.map(t => `"${t}"`).join(', ')}]`
      : `tags: []`;
    content = content.replace(/^tags:\s*\[.+\]/m, newInline);
  }

  writeFileSync(file, content, 'utf-8');
  modifiedCount++;
  const rel = file.replace(contentDir + '/', '');
  const removed = tags.filter(t => removeSet.has(t));
  console.log(`  ${rel}: removed [${removed.join(', ')}]`);
}

console.log(`\nDone. Modified ${modifiedCount} files.`);
